### PR TITLE
query API client impl

### PR DESCRIPTION
**Pull request by yiminc**
this CR does not include encoding stuff, which will be addressed in a separate CR shortly after.

### CODE REVIEW

File: internal_event_handlers.go [hunk 1]
```diff
@@ -39,6 +39,8 @@ import (
 var _ workflowEnvironment = (*workflowEnvironmentImpl)(nil)
 var _ workflowExecutionEventHandler = (*workflowExecutionEventHandlerImpl)(nil)
 
+const queryTypeStackTrace string = "__stack_trace"
```
Comment (mfateev) [inline internal_event_handlers.go]: I think we can make constants for stack trace (and other built in queries) public

File: workflow.go [hunk 2]
```diff
@@ -625,3 +626,43 @@ const DefaultVersion Version = -1
 func GetVersion(ctx Context, changeID string, minSupported, maxSupported Version) Version {
 	return getWorkflowEnvironment(ctx).GetVersion(changeID, minSupported, maxSupported)
 }
+
+// SetQueryHandler sets the query handler to handle workflow query. The queryType specify which query type this handler
```
Comment (mfateev) [inline workflow.go]: I believe that we discussed that query function is executed outside of dispatcher context. We have somehow document that it cannot use any workflow blocking functions like channel.get, etc.

File: workflow.go [hunk 3]
```diff
@@ -625,3 +626,43 @@ const DefaultVersion Version = -1
...
+// All the input parameter must be serializable. You should call cadence.SetQueryHandler() at the beginning of the workflow
+// code.
+// Example of workflow code that support query type "current_state":
+//  func MyWorkflow(ctx cadence.Context, input string) error {
+//    currentState := "started" // this could be any serializable struct
+//    err := cadence.SetQueryHandler(ctx, "current_state", func() (string, error) {
+//      return currentState, nil
+//    })
+//    if err != nil {
+//      currentState = "failed to register query handler"
```
Comment (mfateev) [inline workflow.go]: Should we support querying closed and failed workflows? And what is the semantic of it?
If not this example doesn't make sense as it updates currentState before closing the workflow.

File: internal_event_handlers.go (hunk 1 above)
Comment (yiminc) [reply internal_event_handlers.go]: will change to public and move to Client.go

File: workflow.go (hunk 2 above)
Comment (yiminc) [reply workflow.go]: Will update comments to point that out. However, if user do try to use Channel.Get, the code will block indefinitely. We probably need to error out in that case, meaning that we need to detect if those function is called outside of a workflow context. I will address that in a separate CR.

File: workflow.go (hunk 3 above)
Comment (yiminc) [reply workflow.go]: Technically, we could support. The current implementation does support it. The use case is that user don't need to check the open/closed state of a workflow before sending out query. User could also use query to return workflow result or activity result if they wish to. (However, extending visibility API to return workflow result would be much cheaper.) I don't know, I don't feel strong about this. We can discuss again on this.

**Pull request merged**